### ISSUE TITLE
Soldier and weapon DAE load times

### ISSUE THREAD

**Issue by Bloodknight** [GarageGames/Torque3D#2085]
As part of my now probably tiresome ranting mutterings about the DAE files supplied with the full template, I ran a test.
In debug, it took a full 3 mins 50 seconds to load the empty mission.
In release, it took 20 seconds.

@dottools explained that the system was due to the possibility of the changing DTS format and making sure that the internal assets were up to date with that format

While this particular reasoning I believe requires some other thought, it made me wonder exactly what the debug build is doing that the release isn't with this data especially since the time taken is over 10x as long. If somebody had a project that had say 10x as many assets (not really a stretch with even an indie game) or even more; then the time it would take to regenerate dts files from dae files would be astronomical in debug builds. While I understand that this is a rarity and in real terms should only need to be run once but still, its a potential huge wait.

**Comment by Duion** [comment]
DAE's are not loaded they are converted into .DTS which then are loaded, the conversion takes so long.
It does not matter much how long it takes since you usually just have to do it once for your entire projects lifetime.
You also don't have to use debug mode for that and even if, you usually debug one thing at a time, then the loading time is almost nothing.

**Comment by Bloodknight** [comment]
The first comment is simply semantic nitpicking, try to stop letting your personal biases interfere with actual engine discussions.

The second comment is by the by, potential bugs/issues are still bugs/issues regardless of alternative use cases.

IF this is a one-time-ever scenario, then there should be an external conversion system and the .dae (SOURCE FILES) should not be hard-coded into game scripts. If there is a good reason for deleting dts files and having them reconverted than this issue is an important one.

Try to remember that your development methods are your development methods, everyone else probably has different ways of working. It's probably not uncommon to change 20-100 assets at a time in a larger or more graphically inclined game.

IT could be argued that its bad enough we are stuck with Collada ONLY as it is, or that we are stuck without any real tools to convert files to the torque internal model format. Perhaps all of these problems will be address as an when assimp is fully implemented and we can use whatever formats we want to use, perhaps the Collada converter debug code is deeply flawed and assimp will repair this issue at the same time. 

Personally, I would like to see source art kept a long way from the actual game, it's just an extra pile of work needed to be done as an when a game is packaged up to be shipped.

**Comment by Duion** [comment]
You do not update 100 assets at a time, you create an asset then load it into the engine and the coversion is pretty much instant, maybe a fraction of a second conversion time.
If you load like 100 at once it may add up to the 20 seconds you mentioned and I cannot see any reason to convert them in debug mode which takes that much longer.
And there do exist tools to directly convert to the internal format, but it will not save you any conversion processing time, unless you insist on doing everything in debug.

**Comment by Areloch** [comment]
@Bloodknight Yeah, the main reason to avoid binary files on git is they don't play nice with versioning, so if you make any change, you update the whole file. Also what @dottools said with keeping up with format changes.

This does validly raise the question of how to approach source art with the templates and stuff though. The asset system currently is being oriented towards importing/loading from an external directory(though right now in the R&D it copies to the local folders before processing 'em), but it does keep a reference back to the original path for re-importing/reloading. If we figure on keeping source art files out of the end directories in most cases and only leave it to be imported and use the resulting compiled formats, would it make sense to have the art files moved to a different directory we can source from? 

I suppose one approach is we could have the source art be in the templates dir in it's own folder and have the useable art in the templates be binaries? So when you install/copy the template over on setup, you can launch right away without a delay.

Could lead to some subversioning oddness if we make changes as that'd bulk up the history, but we don't make changes to the art THAT often anyways, so... hm. 

Thoughts?

**Comment by Duion** [comment]
Come on don't try to solve this non-issue, it is totally ridiculous to complain that you have a 20 seconds loading time once you setup a new template compared to like 5-10 or whatever seconds it normally takes to load.

**Comment by John3** [comment]
Hmmm Yes maybe its a potential huge wait.

Nevertheless I think it is better that the effort will be focused on the hiccups. I have a T3D project with ALL the art asset (from Ron, GG and some other folks) released for Torque3D in a gray plane side by side. Never saw an issue in load times, 

... after the load, the hiccups are obvious.....

About the arts source files, a little scripts can do this, move all art source files or delete all of then if needed for release. If needed put at frond-end to this. Retain the same hierarchy of folders, having the orig source file together of dts file, not change this.

FYI BTW right now you can export directly to DTS format using the latest Blender3D... But also what Nathan/dottools said about  DTS format, the actual new version of the plugin export to DTS for Blender3D have an old DTS format, but T3D load without an issue, not sure if something is missing here.

Comment aside, I don't think COLLADA is a big issue regardless I know is necessary to support other formats, I think the problem is how different applications export/import differently to COLLADA, maybe having an ally like the "godot better collada exporter" for Blender3D or having our own plugin :D

**Comment by Areloch** [comment]
There's absolutely nothing wrong with discussing things that could be improved. That's literally the point of the issues section, lol.

John: I agree that the 'first time view' stutters as stuff actually bumps to vram are a problem as well.

**Comment by Duion** [comment]
@John3 
He is not even talking about anything you pointed out, he just talks about the conversion process you get the first time you load up the full template that converts all .dae into .dts.
@Areloch 
But here is nothing that could be improved, converting one asset already takes a fraction of a second, what you want to do? Improve it that it loads in 0.27 seconds instead of 0.3 ?
Bloodknight is just doing useless things in the worst possible setup and then complains, that doing something useless in a bad setup takes longer.

**Comment by Duion** [comment]
How did 20 seconds and 3 minutes 50 seconds now turn into 5 and 15 minutes?
I never had issues with the load times I would rather investigate occasional crashes while loading first time, if they are still there.
